File: 2019/aoc1.py
"""
Advent of Code day 1: The Rocket Equation
"""


def get_fuel(module_mass):
    return module_mass // 3 - 2


def get_recursive_fuel(module_mass):
    fuel = get_fuel(module_mass)
    m = get_fuel(fuel)
    while m > 0:
        fuel += m
        m = get_fuel(m)
    return fuel


with open("data/1.txt") as f:
    lines = f.readlines()


def part_1():
    return sum(get_fuel(int(line)) for line in lines)


def part_2():
    return sum(get_recursive_fuel(int(line)) for line in lines)


if __name__ == "__main__":
    print(part_1())
    print(part_2())

File: 2019/rust/src/aoc1.rs
///
/// Advent of Code day 1: The Rocket Equation
///

//use std::fs;

fn get_fuel(module_mass: &usize) -> usize {
    (module_mass / 3).saturating_sub(2)
}

fn get_recursive_fuel(module_mass: &usize) -> usize {
    let mut fuel = get_fuel(module_mass);
    let mut m = get_fuel(&fuel);
    while m > 0 {
        fuel += m;
        m = get_fuel(&m);
    }
    fuel
}

pub fn get_input() -> Vec<usize> {
    let input = include_str!("../../data/1.txt");
    input.lines().map(|l| l.parse::<usize>().unwrap()).collect()
}

pub fn part_1(masses: &[usize]) -> usize {
    masses.iter().map(get_fuel).sum()
}

pub fn part_2(masses: &[usize]) -> usize {
    masses.iter().map(get_recursive_fuel).sum()
}
